Add type promotion support in sparse tensors

nairbv opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26051 Add type promotion support in sparse tensors**
* #26655 Add torch.promote_types function

Differential Revision: [D17556195](https://our.internmc.facebook.com/intern/diff/D17556195)

nairbv committed:
Add type promotion support in sparse tensors

nairbv committed:
Update on "Add type promotion support in sparse tensors"




[ghstack-poisoned]

nairbv committed:
Update on "Add type promotion support in sparse tensors"




[ghstack-poisoned]

nairbv committed:
Update on "Add type promotion support in sparse tensors"




[ghstack-poisoned]

nairbv committed:
Update on "Add type promotion support in sparse tensors"




[ghstack-poisoned]

gchanan commented:
@ezyang do you think this would cause any problems with your multiple dispatch patches?  I'm thinking the cases of mixed sparse/strided matrices.

gchanan commented on test/test_type_promotion.py:
```diff
@@ -225,6 +225,56 @@ def test_result_type(self):
         self.assertEqual(torch.result_type(torch.tensor(1., dtype=torch.float), torch.tensor(1, dtype=torch.double)), torch.double)
 
+    def test_sparse(self):
+        t = torch.ones([5, 5], dtype=torch.int, device=self.device) * 5
```
nit: you can use `torch.full` instead.

gchanan commented on test/test_type_promotion.py:
```diff
@@ -225,6 +225,56 @@ def test_result_type(self):
+        s = t.to_sparse()
+        self.assertEqual((s + s).to_dense(), t + t)
+        self.assertEqual(t + s, (s + s).to_dense())
+        self.assertEqual(t + s.to(torch.int), (s + s).to_dense())
```
what does `to(int)` do here?

nairbv replied:
leftover from ad-hoc testing, removed

gchanan commented on test/test_type_promotion.py:
```diff
@@ -225,6 +225,56 @@ def test_result_type(self):
+        self.assertEqual((s + s).to_dense(), t + t)
+        self.assertEqual(t + s, (s + s).to_dense())
+        self.assertEqual(t + s.to(torch.int), (s + s).to_dense())
+        self.assertEqual(t + s.to(torch.double), (t + t).to(torch.double))
```
does assertEqual check the dtype?  I know some version did/didn't.

nairbv replied:
it doesn't. I'll add a few more explicit dtype checks.

gchanan commented on test/test_type_promotion.py:
```diff
@@ -225,6 +225,56 @@ def test_result_type(self):
+        self.assertEqual(t + s.to(torch.double), (t + t).to(torch.double))
+        self.assertEqual((s + s.to(torch.double)).to_dense(), t + t)
+
+        inplace = s.clone().to(torch.float)
```
nit: you can just pass `copy=True` to `to`.

gchanan commented on aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:
```diff
@@ -433,9 +454,19 @@ SparseTensor& mul_out_sparse_cuda(SparseTensor& r_, const SparseTensor& t_, cons
+  auto promotedType = promoteTypes(t.scalar_type(), src.scalar_type());
+  TORCH_CHECK(canCast(promotedType, r_.scalar_type()), "Can't convert result type ", promotedType, " to output ", r_.scalar_type());
+
+  if (s_values_.scalar_type() != promotedType) {
```
err..what happens if you did all of this with s instead of s_values?  Would it copy the indices tensor?

nairbv replied:
I think it would yeah. I think it'd also be a bit messier converting src directly since below still needs the values/indices separately, and src is a reference, so there'd be an additional temporary variable that isn't really needed.

ezyang commented:
Yes, this will conflict. I've been trying to land multiple dispatch since all day yesterday. I can give advice where code will have moved afterwards, but all in all it should be pretty straightforward.

nairbv committed:
Update on "Add type promotion support in sparse tensors"




[ghstack-poisoned]

gchanan commented:
@ezyang: just to clarify, you are suggesting we get your patch in first, correct?

ezyang commented:
Yes, that would be preferred!

nairbv committed:
Update on "Add type promotion support in sparse tensors"




[ghstack-poisoned]

nairbv committed:
Update on "Add type promotion support in sparse tensors"




[ghstack-poisoned]

nairbv committed:
Update on "Add type promotion support in sparse tensors"




[ghstack-poisoned]

nairbv commented:
@pytorchbot retest this please

nairbv committed:
Update on "Add type promotion support in sparse tensors"




[ghstack-poisoned]

nairbv committed:
Update on "Add type promotion support in sparse tensors"




[ghstack-poisoned]

nairbv committed:
Update on "Add type promotion support in sparse tensors"




[ghstack-poisoned]

nairbv committed:
Update on "Add type promotion support in sparse tensors"




[ghstack-poisoned]

nairbv committed:
Update on "Add type promotion support in sparse tensors"




[ghstack-poisoned]

gchanan commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -153,7 +153,8 @@ SparseTensor pow_sparse_scalar(const SparseTensor& t, Scalar value) {
 
 Tensor div_sparse(const Tensor& self, const Tensor& value) {
-  Tensor result = at::empty({0}, self.options());
+  auto promotedType = at::result_type(self, value);
```
it doesn't seem like a great idea to call this promotedType when there is a function called promote_types that can return something different than result_type.

nairbv replied:
hmm... result_type does call promote_types though, just that it does so while considering "category." I'm hesitant to call the type resultType because in other contexts it can be cast back into the type of a specified result object, and it seems more readable to be consistent across these functions. This is the intermediate highest-precision promoted type where operations are performed.

E.g. if I do float_tensor *= double_tensor the promoted type is double and operations are performed on doubles, but the result will still be float. The name of the result_type function itself comes from numpy.

nairbv replied:
changed to `commonDtype` since that's equivalent to what we have in TensorIterator.cpp. Does that work for you?

nairbv committed:
Update on "Add type promotion support in sparse tensors"


Differential Revision: [D17556195](https://our.internmc.facebook.com/intern/diff/D17556195)

[ghstack-poisoned]

gchanan reviewed (commented):
rename looks fine.  @zou3519 can you take a look?

nairbv committed:
Update on "Add type promotion support in sparse tensors"


Differential Revision: [D17556195](https://our.internmc.facebook.com/intern/diff/D17556195)

[ghstack-poisoned]

nairbv commented on test/test_type_promotion.py:
```diff
@@ -227,6 +227,59 @@ def test_result_type(self):
+        self.assertEqual((s * 2.0).to_dense().dtype, (t * 2.0).dtype)
+
+        # sparse division only supports division by a scalar
+        self.assertEqual((s / 2).to_dense(), t / 2)
```
added out test for div

nairbv commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -240,6 +242,10 @@ SparseTensor& add_out_sparse_cpu(SparseTensor& r, const SparseTensor& t, const S
   TORCH_CHECK(!r.is_cuda(), "add: expected 'out' to be CPU tensor, but got CUDA tensor");
   TORCH_CHECK(!src.is_cuda(), "add: expected 'other' to be a CPU tensor, but got a CUDA tensor");
 
+  auto commonDtype = promoteTypes(t.scalar_type(), src.scalar_type());
```
ah I didn't realize it was possible to construct a zero-dim sparse tensor.

zou3519 replied:
It is, but on the other hand, I'm not sure who does

nairbv replied:
actually, just tried this, we don't support adding dimensioned tensors to zero-dim tensors. Current implementation of sparse tensors requires t.dims == src.dims() so as long as the dims match promoteTypes will be correct anyway. We should probably expand sparse capabilities but that'll be another issue.

zou3519 replied:
Cool, then there should be an assert here before the `promoteTypes` call so it's clear that we're assuming the inputs are in the same category

zou3519 commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -261,10 +267,22 @@ SparseTensor& add_out_sparse_cpu(SparseTensor& r, const SparseTensor& t, const S
-    get_sparse_impl(r)->set_indices_and_values_unsafe(r_indices, r_values);
+
+    if (commonDtype != r_values.scalar_type()) {
+      r_values = r_values.to(commonDtype);
```
We can speed this up by allocating `r_values = torch.empty_like(r_values)`. We don't actually care about the data that is initially in the result tensor.

zou3519 commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -326,6 +344,10 @@ SparseTensor& add_out_sparse_cpu(SparseTensor& r, const SparseTensor& t, const S
     );
 
+    if (r.scalar_type() != commonDtype) {
+      r_values = r_values.to(r.scalar_type());
```
Instead of doing a `.to`, we can save memory here by doing a copy from `r_values` directly into the values tensor instead of creating a new tensor here and assigning it.

zou3519 replied:
I guess the values tensor of the result might not be large enough to hold all of that...

nairbv replied:
I think for this one we may still need the `to`, otherwise we would have to create a new empty tensor first to copy values into anyway, and I assume that's just basically the implementation of `to`? 

`r_values` here is the promoted version (e.g. float += double, r_values is a double tensor, but we'd need to set r._values to a yet-uncreated float tensor containing those values.

zou3519 replied:
`r` is the result tensor. It's possible that `r` already has data and `r._values()` and `r._indices()` are tensors with some non-trivial storage. If that is the case, instead of discarding values and indices in favor of the new values, we could copy r_values directly into `r._values()`, assuming there is enough memory in `r._values()`.

zou3519 commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -409,10 +437,17 @@ Tensor& add_out_dense_sparse_cpu(Tensor& r, const Tensor& dense, const SparseTen
         });
+    if (r.scalar_type() != commonDtype) {
+      r.copy_(promoted_result);
+    }
```
The implementation is not the nicest; it looks like to do type promotion, the current implementation adds the logic to each possible pathway. Instead of doing that, we could imagine writing a wrapper function around the implementation that promotes the input sparse tensors to the correct dtype, calls the implementation (that operates on tensors of the same dtype), and then copy the result back to the result tensor.

nairbv replied:
yeah I'd like to clean this up eventually, but in doing so I think I'd want to extend support for more combinations of tensors too. Right now arithmetic ops on sparse tensors are pretty limited and didn't want to try to make it generic without actually making the availability of sparse ops more generic.

zou3519 replied:
I don't see why "adding more support for sparse ops" and "making the implementation more generic" have to go together. Right now, the implementation for `add_sparse_out_cpu` is hard to review because there are two code paths that need to be considered, one for where values are contiguous, and one for when they're not.

It would make it easier to review (and also make the implementation look better) if we did something like the following (pseudocode)

```
add_out_sparse_cpu(result, tensor, other) {
  // do type promotion
  common_dtype = result_type(tensor, other)
  result_tmp = empty_like(result, common_dtype)
  tensor_tmp = tensor.to(common_dtype)
  other_tmp = other.to(common_dtype)

  // call the base op
  add_out_sparse_cpu_no_type_promotion(result_tmp, other_tmp, tensor_tmp)

  // shove the result back into its original type.
  result.set_inplace(result_tmp.indices(), result_tmp.values().to(result.dtype), result_tmp.sizes())
}
```

zou3519 replied:
In general putting the type promotion code in helper functions or a wrapper function like this would help because the body of `add_out_sparse_cpu` as well as the other functions are really, really long

zou3519 commented on test/test_type_promotion.py:
```diff
@@ -312,6 +312,64 @@ def test_lt_with_type_promotion(self):
             self.assertTrue(actual, expected)
             self.assertTrue(actual.dtype == torch.bool)
 
+    def test_sparse(self):
```
If I'm reading the code correctly, for add and mul, type promotion is implemented for the following:
- add_out_sparse (cpu + cuda)
- add_out_sparse_dense_sparse (cpu only)
- mul_out_sparse (cpu + cuda)

Why is there a discrepancy in what is supported in `add_out_sparse_dense_sparse`?

zou3519 commented on test/test_type_promotion.py:
```diff
@@ -312,6 +312,64 @@ def test_lt_with_type_promotion(self):
+        self.assertEqual((s + s.to(torch.double)).to_dense(), t + t)
+        self.assertEqual((s + s.to(torch.double)).dtype, torch.double)
+
+        inplace = s.to(torch.float, copy=True)
```
There are a lot of test cases here. I imagine when we add type promotion to more sparse ops, we will not want to just append LOC to this test case (`test_sparse`). Can we split this up into semantically meaningful tests for different ops and different overloads?

ezyang commented:
FYI: this was fake merged due to a bug in ghstack
